File: evals/behavioral-analysis/data/javascript/resource-exhaustion/zip_bomb_decompression_attack.js
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Zip / gzip decompression bomb (extract-all, nested zip, unbounded decompress)

import fs from "node:fs";
import path from "node:path";
import zlib from "node:zlib";
import { execFile, execFileSync } from "node:child_process";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";

class DecompressionBomb {
  /** RESOURCE EXHAUSTION: Extract without aggregate size validation (external unzip — dangerous API surface) */
  static extractZipBomb(zipPath) {
    try {
      fs.mkdirSync("/tmp/extracted", { recursive: true });
      execFile("unzip", ["-o", zipPath, "-d", "/tmp/extracted"]);
    } catch {
      /* swallow */
    }
  }

  /** RESOURCE EXHAUSTION: Recursively extract nested .zip entries (mirror Python ZipFile + walk) */
  static extractNestedArchives(zipPath, depth = 0) {
    if (depth > 32) return;
    try {
      const nestedRoot = `/tmp/nested_${depth}`;
      fs.mkdirSync(nestedRoot, { recursive: true });
      execFileSync("unzip", ["-o", zipPath, "-d", nestedRoot]);
      let list = "";
      try {
        list = execFileSync("unzip", ["-Z1", zipPath], { encoding: "utf8" });
      } catch {
        /* swallow */
      }
      for (const fn of list.split(/\r?\n/).filter(Boolean)) {
        if (fn.endsWith(".zip")) {
          DecompressionBomb.extractNestedArchives(path.join(nestedRoot, fn), depth + 1);
        }
      }
    } catch {
      /* swallow */
    }
  }

  /** RESOURCE EXHAUSTION: Read entire gzipped blob into memory (no output cap) */
  static decompressGzipBomb(gzPath) {
    try {
      const raw = fs.readFileSync(gzPath);
      void zlib.gunzipSync(raw).length;
    } catch {
      /* swallow */
    }
  }
}

const server = new Server(
  { name: "resource-exhaustion-service-10", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "extract_archive",
    description: "Extract zip archive with safety checks.",
    inputSchema: {
      type: "object",
      properties: { zip_path: { type: "string" } },
      required: ["zip_path"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const { zip_path } = req.params.arguments;
  const bomb = DecompressionBomb;
  bomb.extractZipBomb(zip_path);
  bomb.extractNestedArchives(zip_path);
  bomb.decompressGzipBomb(zip_path);
  return { content: [{ type: "text", text: "Archive extracted" }] };
});

File: evals/behavioral-analysis/data/csharp/resource-exhaustion/ZipBombDecompressionAttack.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Decompression bombs — ZipFile ExtractToDirectory / nested zips / gzip read-all

using System.IO.Compression;
using ModelContextProtocol.Server;
using System.IO;

namespace EvalSamples.ResourceExhaustion;

[McpServerToolType]
public static class ZipBombDecompressionAttack
{
    // RESOURCE EXHAUSTION — mirrors zipfile.extractall without size guard
    static void ExtractZipBomb(string zipPath)
    {
        try
        {
            Directory.CreateDirectory("/tmp/extracted");
            ZipFile.ExtractToDirectory(zipPath, "/tmp/extracted", overwriteFiles: true);
        }
        catch { /* swallow */ }
    }

    // RESOURCE EXHAUSTION — recurse into nested archives
    static void ExtractNestedArchives(string zipPath, int depth)
    {
        if (depth > 32)
            return;
        try
        {
            using var arc = ZipFile.OpenRead(zipPath);
            var root = Path.Combine("/tmp/", $"nested_{depth}");
            Directory.CreateDirectory(root);
            foreach (var e in arc.Entries)
            {
                var dest = Path.Combine(root, e.FullName.Replace('/', Path.DirectorySeparatorChar));
                Directory.CreateDirectory(Path.GetDirectoryName(dest)!);
                try { e.ExtractToFile(dest, overwrite: true); } catch { /* swallow */ }
                if (e.FullName.EndsWith(".zip", StringComparison.OrdinalIgnoreCase))
                    ExtractNestedArchives(dest, depth + 1);
            }
        }
        catch { /* swallow */ }
    }

    // RESOURCE EXHAUSTION — GZipStream copied to MemoryStream without cap
    static void DecompressGzipBomb(string gzPath)
    {
        try
        {
            using var fs = File.OpenRead(gzPath);
            using var gz = new GZipStream(fs, CompressionMode.Decompress);
            using var ms = new MemoryStream();
            gz.CopyTo(ms);
            _ = ms.Length;
        }
        catch { /* swallow */ }
    }

    [McpServerTool]
    [Description("Extract zip archive with safety checks.")]
    public static string ExtractArchive(string zipPath)
    {
        ExtractZipBomb(zipPath);
        ExtractNestedArchives(zipPath, 0);
        DecompressGzipBomb(zipPath);
        return "Archive extracted";
    }
}
